### PR TITLE
Document that findnext(f,x,i) and findprev(f,x,i) work for strings.

### PR DESCRIPTION
**Pull request by JamesWrigley**
I think it makes sense to document these, they're quite useful.

### CODE REVIEW

**Comment by JamesWrigley** [comment]
I don't know what's up with the tests, AFAICT they look unrelated :thinking:

**Comment by stevengj** [comment]
Isn't this redundant with the `findnext(predicate::Function, A, i)` docstring here: https://github.com/JuliaLang/julia/blob/e7e8b8900d471f38f90d7150c5123c503aa1ac36/base/array.jl#L2294-L2300 — maybe we should just clarify that docstring to make it clear that that it applies to arbitrary indexed collections (supporting `getindex`, `keys`, and `nextind`).

File: base/strings/search.jl [hunk 1]
```diff
@@ -158,15 +158,26 @@ findfirst(pattern::AbstractVector{<:Union{Int8,UInt8}},
           A::AbstractVector{<:Union{Int8,UInt8}}) =
     _search(A, pattern, firstindex(A))
 
-# AbstractString implementation of the generic findnext interface
-function findnext(testf::Function, s::AbstractString, i::Integer)
-    i = Int(i)
+"""
+    findnext(predicate::Function, s::AbstractString, start::Integer)
+
+Returns the index `i` of the first character after `start` for which
```
Comment (stevengj) [inline base/strings/search.jl:164]:
```suggestion
Returns the index `i` of the first character whose index is `≥ start` for which
```

**Comment by JamesWrigley** [comment]
> Isn't this redundant with the `findnext(predicate::Function, A, i)` docstring here:

That's true, but to me strings and arrays are morally different enough that it makes sense for them to be documented separately. In general I always prefer to document methods (with specializations, like `AbstractString` in this case) for the sake of explicitness for the user to know that such an operation is supported, plus it makes it easier to document any differences in behaviour (though I don't think there's any difference in this case).

I realize that's a personal preference though, I'm happy to move the documentation into the generic `findnext()`/`findprev()` docstring if that's better.

File: base/strings/search.jl (hunk 1 above)
Comment (JamesWrigley) [reply base/strings/search.jl]: Oops, good catch :eyes: Fixed in 445deeaf3b (along with `findprev()`).

**Comment by JamesWrigley** [comment]
(gentle bump)

**Comment by stevengj** [comment]
Since the generic docstring is correct for `String`, I'm reluctant to have a separate docstring saying "it works for `String` too".   Maybe just add a `String` example to the general docstring.

The basic reason is that it makes the documentation harder to read.  When you do `?findnext`, it's hard to read if you have to scroll through multiple docstrings when one would suffice.

**Comment by JamesWrigley** [comment]
Ok, that's fair :+1: I added string examples to the general docstring and mentioned the interface requirements in 62cda7c6ae.

**Review by stevengj — APPROVED** [review]

File: base/array.jl [hunk 2]
```diff
@@ -2294,7 +2294,9 @@ findfirst(A::AbstractArray) = findnext(A, first(keys(A)))
     findnext(predicate::Function, A, i)
 
 Find the next index after or including `i` of an element of `A`
-for which `predicate` returns `true`, or `nothing` if not found.
+for which `predicate` returns `true`, or `nothing` if not found. This works for
+any collection that supports [`getindex`](@ref), [`keys(A)`](@ref), and
+[`nextind`](@ref).
```
Comment (LilithHafner) [inline base/array.jl:2299]: I don't think it's a good idea to say "this works for anything that X" unless X is a documented interface. For example, `findnext` does not work for the type `OneToInf` defined below even though it meets the requirements posed in this PR's docstring.

```julia
julia> struct OneToInf end

julia> function Base.getindex(oti::OneToInf, i)
           @boundscheck i > 0 || throw(BoundsError(oti, i))
           i
       end

julia> Base.keys(oti::OneToInf) = oti

julia> Base.nextind(::OneToInf, i, n=1) = i+n

julia> findnext(==(7), OneToInf(), 5)
ERROR: MethodError: no method matching iterate(::OneToInf)

Closest candidates are:
  iterate(::Base.AsyncGenerator, ::Base.AsyncGeneratorState)
   @ Base asyncmap.jl:362
  iterate(::Base.AsyncGenerator)
   @ Base asyncmap.jl:362
  iterate(::Pkg.Resolve.NodePerm, Any...)
   @ Pkg ~/.julia/juliaup/julia-1.10.0+0.aarch64.linux.gnu/share/julia/stdlib/v1.10/Pkg/src/Resolve/maxsum.jl:240
  ...

Stacktrace:
 [1] first(itr::OneToInf)
   @ Base ./abstractarray.jl:469
 [2] findnext(testf::Base.Fix2{typeof(==), Int64}, A::OneToInf, start::Int64)
   @ Base ./array.jl:2151
 [3] top-level scope
   @ REPL[84]:1

julia> Base.iterate(::OneToInf, i=0) = (i+1, i+1)

julia> findnext(==(7), OneToInf(), 5)
ERROR: MethodError: no method matching lastindex(::OneToInf)

Closest candidates are:
  lastindex(::Any, ::Any)
   @ Base abstractarray.jl:424
  lastindex(::LaTeXStrings.LaTeXString)
   @ LaTeXStrings ~/.julia/packages/LaTeXStrings/ZtSdh/src/LaTeXStrings.jl:109
  lastindex(::Base.JuliaSyntax.SourceFile)
   @ Base /cache/build/default-armageddon-5/julialang/julia-release-1-dot-10/base/JuliaSyntax/src/source_files.jl:132
  ...

Stacktrace:
 [1] last(a::OneToInf)
   @ Base ./abstractarray.jl:524
 [2] findnext(testf::Base.Fix2{typeof(==), Int64}, A::OneToInf, start::Int64)
   @ Base ./array.jl:2152
 [3] top-level scope
   @ REPL[88]:1
```

I would recommend either
- An affirmative list of things it works for;
- Weaken "any" to "most"; or
- Both: "..Arrays, Strings, and most other collections that support..."

Comment (JamesWrigley) [reply base/array.jl]: Right, good point; changed to the last option in 1a4c54c322.

**Comment by LilithHafner** [comment]
Thanks!

**Pull request merged**